sql/src/main/java/io/crate/action/job/ContextPreparer.java

```diff
@@ -88,65 +97,207 @@ public ContextPreparer(MapSideDataCollectOperation collectOperation,
+        }
+    }
+
+    private boolean sameNodeUpstreamDistributionType(UpstreamPhase phase) {
+        return phase.distributionType() == DistributionType.SAME_NODE;
+    }
+
+    public static class PreparerContext {
 
         private final UUID jobId;
-        @Nullable
-        private final RowReceiver rowReceiver;
-        private final NodeOperation nodeOperation;
+        private final Map<Tuple<Integer, Byte>, Boolean> phaseHasSameNodeUpstream = new HashMap<>();
+        private final IntObjectOpenHashMap<NodeOperation> nodeOperations = new IntObjectOpenHashMap<>();
+        private final IntObjectOpenHashMap<RowReceiver> rowReceivers = new IntObjectOpenHashMap<>();
 
-        private PreparerContext(UUID jobId, NodeOperation nodeOperation, @Nullable RowReceiver rowReceiver) {
-            this.nodeOperation = nodeOperation;
+        private final List<ExecutionPhase> executionPhasesToProcess = new ArrayList<>();
+        private RowDownstreamFactory rowDownstreamFactory;
+
+        public PreparerContext(UUID jobId) {
             this.jobId = jobId;
-            this.rowReceiver = rowReceiver;
         }
-    }
 
-    private class InnerPreparer extends ExecutionPhaseVisitor<PreparerContext, ExecutionSubContext> {
+        public boolean phaseHasSameNodeUpstream(int executionPhaseId, byte inputId) {
+            Tuple<Integer, Byte> key = new Tuple<>(executionPhaseId, inputId);
+            Boolean res = phaseHasSameNodeUpstream.get(key);
+            if (res == null) {
+                return false;
+            }
+            return res;
+        }
+
+        public void phaseHasSameNodeUpstream(int executionPhaseId, byte inputId, boolean val) {
+            Tuple<Integer, Byte> key = new Tuple<>(executionPhaseId, inputId);
+            phaseHasSameNodeUpstream.put(key, val);
+        }
 
-        RowReceiver getDownstream(PreparerContext context, UpstreamPhase upstreamPhase, int pageSize) {
-            if (context.rowReceiver == null) {
-                assert context.nodeOperation != null : "nodeOperation shouldn't be null if context.rowDownstream hasn't been set";
-                return rowDownstreamFactory.createDownstream(
-                        context.nodeOperation,
-                        upstreamPhase.distributionType(),
-                        context.jobId,
-                        pageSize);
+        public NodeOperation nodeOperation(int executionPhaseId) {
+            NodeOperation nodeOperation = nodeOperations.get(executionPhaseId);
+            if (nodeOperation == null) {
+                throw new IllegalStateException("NodeOperation not found, must be registered first");
             }
+            return nodeOperation;
+        }
+
+        public void nodeOperation(int executionPhaseId, NodeOperation nodeOperation) {
+            nodeOperations.put(executionPhaseId, nodeOperation);
+        }
+
+        public void rowReceiver(int executionPhaseId, RowReceiver rowReceiver) {
```

Comment:
can this be made private? I think the external caller could use the other method overload that also has the inputId